src/classes/USER_UserService.cls

```diff
@@ -42,34 +42,46 @@ public with sharing class USER_UserService {
     public static final String OBJECT_TYPE_USER = 'User';
 
     /*********************************************************************************************************
-    * @description Determines if a User in the List is a System Administrator based on ProfileId.
-    * @param updatedUsers The User records of the deactivated System Administrators.
-    * @return A Map containing the UserId and Boolean value based on the User's ProfileId.
+    * @description Determines if the Users in the List are "System Administrator" based on assigned permissions;
+    * where a SystemAdmin is defined as a User with Customize Application, Modify All Data, and Author Apex.
+    * @param users The User record(s) to check
+    * @return A Map containing the UserId and Boolean value for HasSystemAdmin Permission
     **********************************************************************************************************/
-    public static Map<Id, Boolean> isSysAdmin(List<User> updatedUsers) {
-        Map<Id, Boolean> userIdToIsSysAdmin = new Map<Id, Boolean>();
-
-        Map<Id, Profile> sysAdminProfiles = new Map<Id, Profile>([SELECT Id
-                , PermissionsCustomizeApplication
-                , PermissionsModifyAllData
-                , PermissionsAuthorApex
-                FROM Profile
+    public static Map<Id, Boolean> isSysAdmin(List<User> users) {
+        Map<Id, Boolean> usersWithAdminPermission = new Map<Id, Boolean>();
+
+        if (users == null || users.isEmpty()) {
+            return usersWithAdminPermission;
+        }
+
+        List<PermissionSetAssignment> permissions = [
+            SELECT Id, AssigneeId
+            FROM PermissionSetAssignment
+            WHERE AssigneeId IN :users
+            AND PermissionSetId IN (
+                SELECT Id
+                FROM PermissionSet
                 WHERE PermissionsCustomizeApplication = TRUE
                 AND PermissionsModifyAllData = TRUE
-                AND PermissionsAuthorApex = TRUE]);
-
-        if (updatedUsers != null && !updatedUsers.isEmpty()) {
-            for (User currentUser : updatedUsers) {
-                userIdToIsSysAdmin.put(currentUser.Id, sysAdminProfiles.containsKey(currentUser.ProfileId) ? true : false);
+                AND PermissionsAuthorApex = TRUE)
+        ];
+        for (User currentUser : users) {
+            Boolean isAdminUser = false;
+            for (PermissionSetAssignment pma : permissions) {
+                if (pma.AssigneeId == currentUser.Id) {
+                    isAdminUser = true;
+                    break;
+                }
             }
+            usersWithAdminPermission.put(currentUser.Id, isAdminUser);
```

Comment:
instead of looping through permissions for every user. Maybe you can use 
`adminAssigneeId = UTIL_SObject.extractId(permissions, PermissionSetAssignment.AssigneeId);`
to create a set. Then on line 69 to 76, it will become
`usersWIthAdminPermission.put(currentUser.Id, adminAssigneeId.contains(currentUserId))`.
This way the permissions list will only need to be loop through once.

Reply:
Great suggestion. Changes made.